### PR TITLE
KX_GameObject kwargs

### PR DESCRIPTION
**Pull request by TryExceptElse**
All KX_GameObject python methods with multiple or optional args can now accept kwargs
test blend: http://pasteall.org/blend/index.php?id=47840

### CODE REVIEW

File: source/gameengine/Ketsji/KX_GameObject.cpp [hunk 1]
```diff
@@ -3759,7 +3759,11 @@ KX_PYMETHODDEF_DOC(KX_GameObject, rayCastTo,
 	const char *propName = "";
 	SCA_LogicManager *logicmgr = GetScene()->GetLogicManager();
 
-	if (!PyArg_ParseTuple(args,"O|fs:rayCastTo", &pyarg, &dist, &propName)) {
+    static const char *kwlist[] = {"other", "dist", "prop", nullptr};
+	if (!PyArg_ParseTupleAndKeywords(
+	        args, kwds, "O|fs:rayCastTo",  const_cast<char**>(kwlist),
+	        &pyarg, &dist, &propName
+	    )) {
```
Comment (panzergame) [inline source/gameengine/Ketsji/KX_GameObject.cpp]: The code style we are using try to put all in one line if possible, if not you have to break and the `{` is not anymore in the same line than the condition but in a new line. Here i think you can put everything in the same line.

File: source/gameengine/Ketsji/KX_GameObject.cpp [hunk 2]
```diff
@@ -4233,8 +4233,14 @@ KX_PYMETHODDEF_DOC(KX_GameObject, addDebugProperty,
 	char *name;
 	int visible = 1;
 
-	if (!PyArg_ParseTuple(args,"s|i:debugProperty", &name , &visible))
+	static const char *kwlist[] = {"name", "debug", nullptr};
+	if (!PyArg_ParseTupleAndKeywords(
```
Comment (panzergame) [inline source/gameengine/Ketsji/KX_GameObject.cpp]: I don't know if this function request keyword, `name` must be provided and after that there's only one value.

File: source/gameengine/Ketsji/KX_GameObject.cpp [hunk 3]
```diff
@@ -3252,12 +3252,16 @@ int KX_GameObject::pyattr_set_lodManager(PyObjectPlus *self_v, const KX_PYATTRIB
...
 }
 
-PyObject *KX_GameObject::PyApplyForce(PyObject *args)
+PyObject *KX_GameObject::PyApplyForce(PyObject *args, PyObject *kwds)
 {
 	int local = 0;
 	PyObject *pyvect;
 
-	if (PyArg_ParseTuple(args, "O|i:applyForce", &pyvect, &local)) {
+    static const char *kwlist[] = {"force", "local", nullptr};
```
Comment (panzergame) [inline source/gameengine/Ketsji/KX_GameObject.cpp]: As for `addDebugProperty` i'm not sure of the interest of making the second argument a keyword. If the doc say the opposite you can change the doc like:
`applyBlah(vector[, local])`

File: source/gameengine/Ketsji/KX_GameObject.cpp [hunk 4]
```diff
@@ -3535,11 +3535,15 @@ PyObject *KX_GameObject::PyApplyImpulse(PyObject *args, PyObject *kwds)
 	return nullptr;
 }
 
-PyObject *KX_GameObject::PySuspendPhysics(PyObject *args)
+PyObject *KX_GameObject::PySuspendPhysics(PyObject *args, PyObject *kwds)
 {
 	int freeConstraints = false;
 
-	if (!PyArg_ParseTuple(args, "|i:suspendPhysics", &freeConstraints)) {
+    static const char *kwlist[] = {"freeConstraints", nullptr};
```
Comment (panzergame) [inline source/gameengine/Ketsji/KX_GameObject.cpp]: Same than `apply**`.

File: source/gameengine/Ketsji/KX_GameObject.cpp [hunk 5]
```diff
@@ -3561,12 +3561,17 @@ PyObject *KX_GameObject::PyRestorePhysics()
 	Py_RETURN_NONE;
 }
 
-PyObject *KX_GameObject::PySuspendDynamics(PyObject *args)
+PyObject *KX_GameObject::PySuspendDynamics(PyObject *args, PyObject *kwds)
 {
 	bool ghost = false;
 
-	if (!PyArg_ParseTuple(args, "|b", &ghost))
+    static const char *kwlist[] = {"ghost", nullptr};
```
Comment (panzergame) [inline source/gameengine/Ketsji/KX_GameObject.cpp]: Same than `apply**`.

File: source/gameengine/Ketsji/KX_GameObject.cpp [hunk 6]
```diff
@@ -1993,7 +1993,11 @@ PyObject *KX_GameObject::PyReinstancePhysicsMesh(PyObject *args)
 	PyObject *gameobj_py= nullptr;
 	PyObject *mesh_py= nullptr;
 
-	if (!PyArg_ParseTuple(args,"|OOi:reinstancePhysicsMesh",&gameobj_py, &mesh_py, &dupli) ||
+    static const char *kwlist[] = {"gameObject", "meshObject", "dupli", nullptr};
```
Comment (panzergame) [inline source/gameengine/Ketsji/KX_GameObject.cpp]: This one is good :).

File: source/gameengine/Ketsji/KX_GameObject.cpp [hunk 7]
```diff
@@ -4125,8 +4125,14 @@ KX_PYMETHODDEF_DOC(KX_GameObject, stopAction,
 {
 	short layer = 0;
 
-	if (!PyArg_ParseTuple(args, "|h:stopAction", &layer))
+    static const char *kwlist[] = {"layer", nullptr};
```
Comment (panzergame) [inline source/gameengine/Ketsji/KX_GameObject.cpp]: Same as `apply**`

File: source/gameengine/Ketsji/KX_GameObject.cpp [hunk 8]
```diff
@@ -4170,8 +4188,14 @@ KX_PYMETHODDEF_DOC(KX_GameObject, setActionFrame,
 	short layer = 0;
 	float frame;
 
-	if (!PyArg_ParseTuple(args, "f|h:setActionFrame", &frame, &layer))
+    static const char *kwlist[] = {"frame", "layer", nullptr};
```
Comment (panzergame) [inline source/gameengine/Ketsji/KX_GameObject.cpp]: Same as `apply**`

**Comment by TryExceptElse** [comment]
removed kwarg parsing from methods that only take 1 or 2 args and updated documentation to reflect which methods take kwargs, and which only accept positional args.

**Comment by TryExceptElse** [comment]
[new test .blend](http://pasteall.org/blend/index.php?id=47843)

File: doc/python_api/rst/bge_types/bge.types.KX_GameObject.rst [hunk 9]
```diff
@@ -595,7 +599,7 @@ base class --- :class:`SCA_IObject`
       :return: The vector in relation to the objects rotation.
       :rtype: 3d vector.
 
-   .. method:: applyMovement(movement, local=False)
+   .. method:: applyMovement(movement[, local])
```
Comment (youle31) [inline doc/python_api/rst/bge_types/bge.types.KX_GameObject.rst]: hello, it seems you forgot to indicate the default local value here and for applyRotation too.

**Comment by youle31** [comment]
For me appart from the remark I noted (in the rst file), it sounds good. I have another small remak but it isn't so important to me and in KX_GameObject, most of the methods have the code style you choosed but usually when we are not in a switch: case or in a function definition, we put the first brackets "{" on the same line than the if. But as other methods in this file don't follow this rule, it's just a remark and ont a request to change that. I didn't tested, but this sounds good to me.

**Comment by youle31** [comment]
For me it's ok. Thanks for your work TryExceptElse

**Comment by panzergame** [comment]
@TryExceptElse : I fixed some doc and source issue in branch ge_kwargs, could you check it please ?

**Pull request closed without merging**